### ISSUE TITLE
Forking fibers with other effects

### ISSUE THREAD

**Issue by fwbrasil** [getkyo/kyo#229]
## Problem

There's an important design challenge with Kyo's flexible effects: forking fibers can't be easily done with arbitrary effects. ZIO for example is able to "hardcode" the effect handling when forking a fiber because both effects it supports, dependency injection and short circuiting, can be unambiguously handled by the runtime when forking a fiber. It only needs to save the environment values from the forking call site and propagate failures correctly. That's not the case with Kyo since the pending effect set is essentially unbounded and can have arbitrary effects. We still don't recommend users creating new effects, but that's likely to become the case for advanced users at some point.

Currently, users have to manually handle other effects before forking a fiber:

```scala
def ko(v: Int < Aborts[String]) = Fibers.init(v) // Compilation failure

def ok(v: Int < (Aborts[String] & IOs)): Fiber[Either[String, Int]] < IOs =
    Fibers.init(Aborts[String].run(v))
```

This approach is confusing and cumbersome. I've been exploring this problem for some time but I couldn't identify a design that seems to have the usability I'm looking for yet. This problem can actually be split into two:

1. **Explicit forking**: Methods that return a `Fiber` instance like `Fibers.init` and `Fibers.parallelFiber`. This is the most difficult case to provide a good UX since there's an expectation that the `Fiber` instance represents a computation that will finalize at some point, not suspend again with some effect.
2. **Implicit forking**: Methods that fork fibers but don't return a `Fiber` instance like `Fibers.parallel` and `Fibers.race`. In these cases where the method immediately joins a fiber are a bit easier because the API doesn't need to provide a `Fiber` representing the end of the computation.

### Forking combinators for larger effects

This has been the main approach for some time now. The `Joins` [trait](https://github.com/getkyo/kyo/blob/e03e9d3317a535b55eea17d33c2e0524bb3d7107/kyo-core/shared/src/main/scala/kyo/Joins.scala#L3) provides forking combinators that can be extended by other effects. The main example of this approach is the `AIs` effect. It extends `Joins` and [implements](https://github.com/getkyo/kyo/blob/main/kyo-llm/shared/src/main/scala/kyo/llm/ais.scala#L213) the correct wiring for its specific underlying effects.

This solution only addresses the "Implicit forking" problem, though. It's still not possible to call `Fibers.init` with a pending `AIs` effect for example. An interesting observation of this approach is that solving only the "Implicit forking" problem might be a viable solution since it covered pretty much all of my usage of the `AIs` effect.

### Fibers with arbitrary pending effects

We could allow fibers to have arbitrary pending effects. The type would become `Fiber[T, S]` and when `fiber.get` is called a `T < (Fibers & S)` would be returned. This approach is relatively easy to implement but I don't think it provides a good mental model. For example, with this approach the following code would run strictly sequentially:

```scala
val a: Fiber[Seq[Int], Envs[Int]] < IOs = Fibers.parallel(Seq.fill(10)(Envs[Int].get.map(_ + 1)))
val b: Seq[Int] < (Fibers & Envs[Int]) = a.map(_.get)
val c: Seq[Int] < Fibers = Envs[Int].run(42)(b)
```

Fibers would initially start in parallel, get immediately suspended with the `Envs` effect, and then `Envs[Int].run` would later handle each of the elements sequentially. This approach doesn't seem to offer a reasonable mental model for how to use fibers.

### Joins

I explored providing a solution that would allow users to express the effect handling order of a computation when forking. The mechanism would take care of performing the steps to handle and suspend effects again transparently. The main abstraction is the `Joins` interface, which allows effects to `save` state from the call site, `handle` effects within the forked fiber, and then call `resume` to suspend the effect again at the end of the fiber execution.

```scala
// minimized API
trait Joins[E] { self =>
  type M[_]
  type State
  
  def save: State < E
  def handle[T, S](s: State, v: T < (E & S)): M[T] < S
  def resume[T, S](v: M[T] < S): T < (E & S)
  def andThen[E2](j: Joins[E2]): Joins[E & E2]
}

// example implementations
implicit def aborts[E: Tag]: Joins[Aborts[E]] =
  new Joins[Aborts[E]] {
    type State = Unit
    type M[T]  = Either[E, T]
    def save = ()
    def handle[T, S](s: State, v: T < (Aborts[E] & S))(implicit f: Flat[T < (Aborts[E] & S)]) =
      Aborts[E].run(v)
    def resume[T, S](v: M[T] < S) =
      Aborts[E].get(v)
  }

implicit def envs[E: Tag]: Joins[Envs[E]] =
  new Joins[Envs[E]] {
    type State = E
    type M[T]  = T
    def save = envs.get
    def handle[T, S](s: State, v: T < (Envs[E] & S))(implicit f: Flat[T < (Envs[E] & S)]) =
      Envs[E].run(s)(v)
    def resume[T, S](v: M[T] < S) =
      v
  }
```

When forking fibers, users would specify the handling order that would summon a `Joins` instance implicitly:

```scala
val a: Int < (Aborts[Exception] & Envs[Int]) = ...
val b: Seq[Int] < (Aborts[Exception] & Envs[Int] & Fibers) = 
  Fibers.parallel(Joins[Aborts[Exception], Envs[Int]])(Seq.fill(10)(v))
```

Something I like about the `Joins` abstraction is that it provides a convenient way to enforce the effect handling order of a computation so it could be useful even in other scenarios. The main issue with this approach is that it also only solves the "Implicit forking" problem and its abstraction level might to be too high for newcomers. I pushed a branch some time ago with this impl: https://github.com/getkyo/kyo/compare/main...joins2.

### Allow IOs-based effects when forking

This was my latest attempt. It's possible to implement effects like `Resources` and `Envs` based only on the `IOs` effect, which is ok to accept when forking fibers. I changed the `Resources` effect to be [based only on IOs](https://github.com/getkyo/kyo/blob/main/kyo-core/shared/src/main/scala/kyo/resources.scala#L5) instead of a separate effect and made `Fibers.init` accept any IOs-based pending effect. With this recent change, it's already possible to fork fibers using `Resources`:

```scala
val a: Int < Resources = ...
val b: Int < (Resources & IOs) = Fibers.init(a)
```

This approach is much more convenient because it solves both the explicit and implicit forking problems. Encoding `Envs` in terms of `Locals`/`IOs` is something I've [explored for performance](https://github.com/getkyo/kyo/compare/main...ios-based-envs) so it might not be very problematic but lowering `Aborts` to `IOs` seems to require introducing something like `Halt` in Kyo's core. I'd also like to avoid having `IOs` everywhere. For example, it doesn't seem ideal that someone working only with `Envs` ends up having to handle `IOs`.

## Conclusion

I'd encourage thinking out-of-the box here since these approaches still seem to have significant downsides. Something I've been considering is keeping `Fibers` as an inflexible low-level effect and then creating a new effect that handles only the "implicit fork" with better UX.

cc/ @kitlangton

**Comment by johnhungerford** [comment]
I agree that it's better not to have effects like `Envs` and `Aborts` based on `IOs`. It's really great that kyo can work with pure values and effects seamlessly. I wouldn't give this up for a more flexible fiber model.

As it seems to me, explicit forking with arbitrary effects just doesn't make sense: if a fiber is understood to be a running effect, it should not be a suspended one. You should, however, be able to fork `Aborts` effects in addition to `IOs`, since an Fiber suspended with a failure can be understood as being "finished."

I therefore think that we should focus on implicit forking. In this case, I would think that something like `Joins` makes sense, only I don't see why the user would have to specify handling order when they call an implicitly forking method like `Fibers.parallel`. Why can't that order be determined by the order in which the entire effect is handled? Effects that implement `Joins` should be able to pass their handlers to forked fibers, handle the running fiber, and then unhandle (i.e. `resume`) the result. What am I missing here?

As I envision it, the API would look something like this:

```scala
trait BaseFiber[+T, -S](using Joinable[S])
    // Same as current Fiber interface except:
    def get: T < (Fibers & S) // instead of T < Fibers
    def getTry: Try[T] < (Fibers & S)
    def toFuture: Future[T < S] < IOs

type Fiber[+T, +E] = BaseFiber[T, Aborts[E]]

object Fibers:
    def fork[T, E, S >: Fibers & Aborts[E]](v: T < (Fibers & S)): Fiber[T, E] < IOs = ???

    def forkJoin[T, S, T1, S1](vs: Seq[T < S])(using Joinable[S])(join: Seq[T < S] => T1 < S1): T1 < (IOs & S1) =
        // Uses BaseFiber under the hood
        ???
```

**Comment by fwbrasil** [comment]
Bounty requirements:

1. Discuss potential designs first
1. We probably need to solve only the "implicit forking" problem
2. New benchmarks to validate the performance of the changes

**Comment by fwbrasil** [comment]
@pablf thank you for looking into this. I'm sorry that I haven't provided feedback, I had seen your analysis before my vacation and forgot to follow up. I've also spent a good time exploring exploring approaches to support any effect, I appreciate your effort :) The trouble is that the effect suspensions escape the scope of the regular effect handling. For example:

```scala
object Fiber:
    def init[T, S](v: T < S): Fiber[T] < S

// Thread A forks a new fiber to execute concurrently, which may suspend with `Env` at any point
val a: Fiber[Int] < (Env[Int] & IO) = Fiber.init(Env.get[Int].map(_ + 1)
// While Thread B might still be running the fiber, Thread A handles the `Env` effect 
val b: Fiber[Int] < IO = Env.run(1)(a)
```

I've worked on a solution for this ticket as part of the library redesign work. The library's new kernel has two kinds of effects: one that functions as an abstract arrow (named `Effect`) and another that propagates information as part of the context of the computation (`ContextEffect`, very similar to `Locals` or ZIO's environment). Propagating context effects to fibers is easy since it's just scoped values. Several other effects are now also context effects: `Env`, `Local`, `Resource`, and other effects based on those like `Clock` and `Random`. 

Besides the context effects, the new design supports `Abort` by adding a new `E` type parameter to `Fiber`. For example, a computation `Int < (Abort[Exception] & Async)` becomes a `Fiber[Exception, Int]` when forked with `Async.run` and `fiber.get` transforms it back to `T < (Abort[Exception] & Async)`.

I'll close this ticket since I believe the new design addresses it but, if you're still interested in contributing, I'd be happy to help you find some other interesting ticket and set up a bounty. Feel free to reach out on discord as well!
